# Pangenomic Identification of Pathogenic Factors

## Goal
Defining the clinically relevant elements of the pangenome of Staphylococcus aureus and other pathogenic bacteria.   Challenge: mobile genetic elements are ambiguous as repeats, phages, plasmids, etc

 Identification of unique accessory elements in individual patient samples
Differentiation between screening (nose swabs) and lesions
Building a rapid pipeline to differentiate between phage, plasmid and low complexity regions in these differential sites.  
Extending that to other bacterial species

## Backgound and Significance
### Staphylococcus aureus genomes and mobile genetic elements (MGE).
 The proliferation of S. aureus genomic sequences in public databases reflects strong interest in understanding S. aureus genome diversity and evolution. With an average of 2,800 coding sequences, it is estimated that 56% of S. aureus genes are shared by all S. aureus strains. These genes constitute the ‘core genome’ in opposition to the ‘accessory genome’, which is variable between strains and mostly made of mobile genetic elements (MGE) enriched in hypothetical proteins. 

 Staphylococcal MGE encompass any intra- or extra-chromosomal DNA segment that can be independently mobilized within or between S. aureus cells. MGE include plasmids, transposons, integrons, genomic islands, S. aureus pathogenicity islands (SaPIs), integrative conjugative elements, staphylococcal chromosome cassettes, and phages. Together, phages and plasmids are the main source of MGE diversity among S. aureus strains. Plasmids – auto-replicating, covalently closed, circular double-stranded DNA molecules – usually contain genes required to sustain environmental challenges including antibiotic exposure and/or immune pressures. Phages are bacterial viruses that can integrate into the S. aureus chromosome. As the primary vehicles for horizontal transfer of MGEs between strains, phages are major drivers of staphylococcal genome evolution. Both, plasmids and phages, encode well characterized virulence genes including toxins or super-antigens but for the most part, the function of their protein content remains elusive. 

### Why study S. aureus MGE diversity?
 MGE discovery and characterization are important goals for clinical genomics analysis because almost all S. aureus strains harbor several MGE with potentially syndrome- and tissue-specific functions. As of August 2017,  327 unique plasmid and 87 phage sequences have been identified and deposited in the US National Center for Biotechnology and Information (NCBI) database. While more MGE are discovered everyday, highlighting the breadth of S. aureus MGE diversity, the extent and importance of MGE contribution to S. aureus pathogenesis is largely under-appreciated. This is mainly due to the complications inherent to their identification, analysis and characterization in vivo. To circumvent this limitation, I aim  to develop and validate a bioinformatic pipeline combining genomics and state of the art algorithms allowing to assess for the presence of any MGE in a specific genome and determine potential associative correlations with S. aureus diseases. 

## WorkFlow
![My image](https://github.com/Rickcopin/Pathogenic_Pangenomes/blob/master/images/approaches.png)
